android/BOINC/app/src/main/java/edu/berkeley/boinc/rpc/RpcClient.java

```diff
@@ -163,12 +168,51 @@ private static String modeName(int mode) {
      * Methods for connection - opening/closing/authorization/status
      */
 
+    /**
+     * Connect to BOINC core client via TCP Socket (abstract, "boinc_socket")
+     *
+     * @return true for success, false for failure
+     */
+    public boolean open(String address, int port) {
+        mIsRemote = true;
+        if (isConnected()) {
+            // Already connected
+            if (Logging.LOGLEVEL <= 4)
+                Log.e(Logging.TAG, "Attempt to connect when already connected");
+            // We better close current connection and reconnect (address/port could be different)
+            close();
+        }
+        try {
+            mTcpSocket = new Socket();
+            mTcpSocket.connect(new InetSocketAddress(address, port), CONNECT_TIMEOUT);
+            mTcpSocket.setSoTimeout(READ_TIMEOUT);
+            socketSource = Okio.buffer(Okio.source(mTcpSocket.getInputStream()));
+            socketSink = Okio.buffer(Okio.sink(mTcpSocket.getOutputStream()));
+        } catch (IllegalArgumentException e) {
+            if (Logging.LOGLEVEL <= 4)
+                Log.e(Logging.TAG, "connect failure: illegal argument", e);
+            mTcpSocket = null;
+            return false;
+        } catch (IOException e) {
+            if (Logging.WARNING) Log.w(Logging.TAG, "connect failure: IO", e);
+            mTcpSocket = null;
+            return false;
+        } catch (Exception e) {
+            if (Logging.WARNING) Log.w(Logging.TAG, "connect failure", e);
+            mTcpSocket = null;
+            return false;
+        }
+        if (Logging.DEBUG) Log.d(Logging.TAG, "Connected successfully");
+        return true;
+    }
+
     /**
      * Connect to BOINC core client via Unix Domain Socket (abstract, "boinc_socket")
      *
      * @return true for success, false for failure
      */
     public boolean open(String socketAddress) {
```

Comment:
Content of
`public boolean open(String address, int port)`
and
`public boolean open(String socketAddress)`
is almost identical. This is not good.
Try either to merge these functions into one with condition check inside or move the common part to the different function that will be called from these functions.

Reply:
To do function with 3 parameter that not all use (and put null in some of them) and choose from the paramter base on remote is a bad design. That why I am against to merge these functions into one.
I try to implement the other option that is better, with a base class of both tcpSocket and localSocket.
But to switch from localSocket to base class need api 17 (we are on 16).
To switch from tcpSocket to base class need api 19.

Reply:
I did it with obj.

Reply:
I refactor and did in other way.